### DISCUSSION TITLE
How do I declare a template of function with `inout` parameter in Cppfront?

### DISCUSSION THREAD

**Discussion by lauler1** [hsutter/cppfront#843] [Q&A]
Hi,

I have a silly question on the syntax in cppfront.

I have a function like: `myfunc: (inout var: Mytype) = { … }`.

I would like to pass this function ` myfunc ` as a parameter to another function using `std::function`.

I tried `anotherfunc: (funcpar: std::function<void(Mytype&)>)` but Cppfront does not accept the `&`.
If I remove the `&` then c++ complains that the types don't match.

 How do I declare a template of function that contains `inout` parameter in Cppfront?

Thanks

**Comment by JohelEGP** [comment]
While we lack support (which #526 aims to remedy),
you can use `std::function<void(std::add_lvalue_reference_t<Mytype>)>`.

**Comment by lauler1** [comment]
Worked!

Thank you very much!

**Comment by hsutter** [comment]
You can also `std::function< decltype(f) >` where `f` is a function with the intended signature.

This works:

```
Mytype: type = { }

myfunc: (inout var: Mytype) = { std::cout << "myfunc"; }

main: () = {
    f: std::function< decltype(myfunc) > = myfunc&;
    m: Mytype = ();
    f(m);
    _ = m;
}
```

**Comment by ntrel** [comment]
> I tried anotherfunc: (funcpar: std::function<void(Mytype&)>) but Cppfront does not accept the &.

(cppfront does but rewrites it to `void(&Mytype)`, causing a Cpp1 error). I'm a bit surprised cppfront even accepts `void(Mytype)` as a template argument, it doesn't allow it in a type alias.

Presumably at some point this will work:
```c++
myfunc: (inout var: Mytype);
anotherfunc: (funcpar: *(inout Mytype) -> void);

anotherfunc(myfunc&);
```

**Reply by JohelEGP** [reply]
> I'm a bit surprised cppfront even accepts `void(Mytype)` as a template argument, it doesn't allow it in a type alias.

Cpp2 parses `void(Mytype)` as an expression.
Cpp1 interprets it as a type template argument.
See the second point of the second list: <https://github.com/hsutter/cppfront/discussions/628#discussioncomment-6867807>.
